tests/holdings_maintenance/readonly_roots.py

```
@@ -0,0 +1,145 @@
+
+ENV_VAR = 'PDSFILE_READONLY_ROOTS'
+
+# open() modes that can create or change a file. 'r' alone is the only one that cannot.
+_WRITE_FLAGS = frozenset('wxa+')
+
+
+class ReadOnlyHoldingsError(PermissionError):
+    """Raised when a test, or a tool a test started, writes into a real holdings root."""
+
+
+def _roots():
```

Ruff (0.16.1)

[warning] Missing return type annotation for private function `_roots`

(ANN202)